pdf/model/appender.go

```diff
@@ -370,20 +380,83 @@ func (a *PdfAppender) ReplacePage(pageNum int, page *PdfPage) {
 	}
 }
 
-// ReplaceAcroForm replaces the acrobat form. It appends a new form to the Pdf which replaces the original acrobat form.
+// Sign signs a specific page with a digital signature using a specified signature handler.
+// Returns an Acroform and PdfAppearance that can be used to customize the signature appearance.
+func (a *PdfAppender) Sign(pageNum int, handler SignatureHandler) (acroForm *PdfAcroForm, appearance *PdfAppearance, err error) {
+	acroForm = a.Reader.AcroForm
+	if acroForm == nil {
+		acroForm = NewPdfAcroForm()
+	}
+	pageIndex := pageNum - 1
+	var page *PdfPage
+	for i, p := range a.pages {
+		if i == pageIndex {
+			page = p.Duplicate()
+			break
+		}
+	}
+	if page == nil {
+		return nil, nil, fmt.Errorf("page %d not found", pageNum)
+	}
+
+	// TODO add more checks before set the fields
+	acroForm.SigFlags = core.MakeInteger(3)
+	acroForm.DA = core.MakeString("/F1 0 Tf 0 g")
+	n2ResourcesFont := core.MakeDict()
+	n2ResourcesFont.Set("F1", DefaultFont().ToPdfObject())
+	acroForm.DR = NewPdfPageResources()
+	acroForm.DR.Font = n2ResourcesFont
+	sig := NewPdfSignature()
+	sig.M = core.MakeString(time.Now().Format("D:20060102150405-07'00'"))
+	//sig.M = core.MakeString("D:20150226112648Z")
+	sig.Type = core.MakeName("Sig")
+	sig.Reason = core.MakeString("Test1")
+	if err := handler.InitSignature(sig); err != nil {
+		return nil, nil, err
+	}
+	a.addNewObjects(sig.container)
+
+	appearance = NewPdfAppearance()
+
+	fields := append(acroForm.AllFields(), appearance.PdfField)
+	acroForm.Fields = &fields
+
+	procPage(page)
+
+	appearance.V = sig.ToPdfObject()
+	appearance.FT = core.MakeName("Sig")
+	appearance.V = sig.ToPdfObject()
+	appearance.T = core.MakeString("Signature1")
+	appearance.F = core.MakeInteger(132)
+	appearance.P = page.ToPdfObject()
+	appearance.Rect = core.MakeArray(core.MakeInteger(0), core.MakeInteger(0), core.MakeInteger(0),
+		core.MakeInteger(0))
+	appearance.Signature = sig
+
+	// Add the signature appearance to the page annotations.
+	page.Annotations = append(page.Annotations, appearance.PdfAnnotationWidget.PdfAnnotation)
+
+	a.pages[pageIndex] = page
+
+	// Update acroform.
+	a.ReplaceAcroForm(acroForm)
+
+	return acroForm, appearance, nil
+}
+
+// ReplaceAcroForm replaces the acrobat form. It appends a new form to the Pdf which
+// replaces the original AcroForm.
 func (a *PdfAppender) ReplaceAcroForm(acroForm *PdfAcroForm) {
 	a.acroForm = acroForm
 }
 
 // Write writes the Appender output to io.Writer.
+// It can only be called once and further invokations will result in an error.
 func (a *PdfAppender) Write(w io.Writer) error {
-	if _, err := a.rs.Seek(0, io.SeekStart); err != nil {
-		return err
-	}
-	offset, err := io.Copy(w, a.rs)
-	if err != nil {
-		return err
+	if a.written {
+		return errors.New("appender write can only be invoked once")
 	}
+	a.written = true
```

Comment:
Should the `written` field be set at the end of the method, after a successful write?